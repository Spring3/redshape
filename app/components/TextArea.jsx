import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const StyledTextArea = styled.textarea`
  font-size: 14px;
  font-family: inherit;
`;

const TextArea = ({
  name,
  onChange,
  rows,
  disabled,
  id,
  value,
  className
}) => (
  <StyledTextArea
    name={name}
    onChange={onChange}
    rows={rows}
    disabled={disabled}
    id={id}
    className={className}
  >
    {value}
  </StyledTextArea>
);

TextArea.propTypes = {
  name: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]),
  onChange: PropTypes.func.isRequired,
  rows: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]),
  disabled: PropTypes.bool,
  id: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]),
  className: PropTypes.string,
  value: PropTypes.string
};

TextArea.defaultProps = {
  name: undefined,
  rows: 2,
  disabled: false,
  id: undefined,
  className: undefined,
  value: undefined
};

export default TextArea;
