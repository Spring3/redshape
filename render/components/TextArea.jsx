import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const StyledTextArea = styled.textarea`
  font-size: 14px;
  resize: none;
  overflow: hidden;
  height: auto;
  font-family: inherit;
  width: 100%;
  line-height: 1.5rem;
  box-sizing: border-box;
`;

const TextArea = ({
  name,
  onChange,
  rows,
  disabled,
  id,
  value,
  className,
  forwardedRef,
  onKeyDown
}) => (
  <StyledTextArea
    ref={forwardedRef}
    name={name}
    onChange={onChange}
    rows={rows}
    disabled={disabled}
    id={id}
    className={className}
    value={value}
    onKeyDown={onKeyDown}
  />
);

TextArea.propTypes = {
  name: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]),
  onChange: PropTypes.func.isRequired,
  rows: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]),
  disabled: PropTypes.bool,
  id: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]),
  className: PropTypes.string,
  value: PropTypes.string,
  forwardedRef: PropTypes.object,
  onKeyDown: PropTypes.func
};

TextArea.defaultProps = {
  name: undefined,
  rows: 1,
  disabled: false,
  id: undefined,
  className: undefined,
  value: undefined,
  forwardedRef: undefined
};

export default React.forwardRef((props, ref) => <TextArea forwardedRef={ref} {...props} />);
