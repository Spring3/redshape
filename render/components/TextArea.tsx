import React, { ChangeEventHandler, FocusEventHandler, forwardRef, KeyboardEventHandler } from 'react';
import { css } from '@emotion/react';
import { useTheme } from 'styled-components';
import { theme as Theme } from '../theme';

const styles = {
  textArea: (theme: typeof Theme) => css`
    font-size: 14px;
    resize: none;
    overflow: hidden;
    height: auto;
    font-family: inherit;
    width: 100%;
    line-height: 1.5rem;
    box-sizing: border-box;
    background: ${theme.bg};
  `,
  disabled: (theme: typeof Theme) => css`
    background: ${theme.bgDisabled};
    border-color: ${theme.bgDarker} !important;
    color: ${theme.minorText};

    &:hover {
      border-color: ${theme.bgDarker} !important;
    }
  `
};

type TextAreaProps = {
  name?: string;
  onChange?: ChangeEventHandler<HTMLTextAreaElement>;
  rows?: number;
  disabled?: boolean;
  id?: string;
  value?: string;
  className?: string;
  maxLength?: number;
  onBlur?: FocusEventHandler<HTMLTextAreaElement>;
  onKeyDown?: KeyboardEventHandler<HTMLTextAreaElement>
}

const TextArea = forwardRef<HTMLTextAreaElement, TextAreaProps>(({
  name,
  onChange,
  onBlur,
  rows = 1,
  disabled = false,
  id,
  value,
  className,
  maxLength,
  onKeyDown
}: TextAreaProps, ref) => {
  const theme = useTheme() as typeof Theme;
  return (
    <textarea
      ref={ref}
      css={[
        styles.textArea(theme),
        disabled ? styles.disabled(theme) : ''
      ]}
      name={name}
      onChange={onChange}
      onBlur={onBlur}
      rows={rows}
      maxLength={maxLength}
      disabled={disabled}
      id={id}
      className={className}
      value={value}
      onKeyDown={onKeyDown}
    />
  );
});

export {
  TextArea
};
