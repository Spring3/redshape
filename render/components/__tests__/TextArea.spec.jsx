import React from 'react';
import { render, fireEvent, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import TextArea from '../TextArea';

describe('TextArea component', () => {
  afterEach(cleanup);

  it('should render the text area', () => {
    render(<TextArea onChange={() => {}} />);
    expect(document.querySelector('textarea')).toBeDefined();
  });

  it('should react to the onChange event', () => {
    const onChange = jest.fn();
    render(
      <TextArea onChange={onChange} />
    );

    const event = {
      target: {
        value: '123test'
      },
      persist: () => {}
    };

    fireEvent.change(document.querySelector('textarea'), event);

    expect(onChange).toHaveBeenCalled();
  });
});
