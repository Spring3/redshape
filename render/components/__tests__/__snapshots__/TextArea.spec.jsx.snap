// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea component should match the snapshot 1`] = `
.c0 {
  font-size: 14px;
  resize: none;
  overflow: hidden;
  height: auto;
  width: 100%;
  line-height: 1.5rem;
  box-sizing: border-box;
}

<textarea
  className="c0"
  disabled={false}
  onChange={[Function]}
  rows={1}
/>
`;
